pint_xarray/checking.py

```diff
@@ -0,0 +1,237 @@
+                   or list of mapping of hashable to unit-like, optional
+        The expected units of the returned value(s), either as a
+        single unit or as a list of units. The decorator will attach
+        these units to the variables returned from the function.
+
+        A value of None indicates not to attach any units to that
+        return value (suitable for flags and other non-data results).
+    kwargs_units : mapping of hashable to unit-like, optional
+        Unit to expect for each keyword argument given to func.
+
+        The decorator will first check that arguments passed to the
+        decorated function possess these specific units (or will
+        attempt to convert the argument to these units), then will
+        strip the units before passing the magnitude to the wrapped
+        function.
+
+        A value of None indicates not to check that argument for units
+        (suitable for flags and other non-data arguments).
+
+    Returns
+    -------
+    return_values : Any
+        Return values of the wrapped function, either a single value
+        or a tuple of values. These will be given units according to
+        return_units.
+
+    Raises
+    ------
+    TypeError
+        If an argument or return value has a specified unit, but is
+        not an xarray.DataArray or pint.Quantity.  Also thrown if any
+        of the units are not a valid type, or if the number of
+        arguments or return values does not match the number of units
+        specified.
+
+    Examples
+    --------
+
+    Decorating a function which takes one quantified input, but
+    returns a non-data value (in this case a boolean).
+
+    >>> @expects("deg C")
+    ... def above_freezing(temp):
+    ...     return temp > 0
+
+    Decorating a function which allows any dimensions for the array, but also
+    accepts an optional `weights` keyword argument, which must be dimensionless.
+
+    >>> @expects(None, weights="dimensionless")
+    ... def mean(da, weights=None):
+    ...     if weights:
+    ...         return da.weighted(weights=weights).mean()
+    ...     else:
+    ...         return da.mean()
+
+    """
+
+    def _expects_decorator(func):
+
+        # check same number of arguments were passed as expected
+        sig = inspect.signature(func)
+        positional_args = (
+            Parameter.POSITIONAL_ONLY,
+            Parameter.VAR_POSITIONAL,
+            Parameter.POSITIONAL_OR_KEYWORD,
+        )
+        n_args = len(
+            [
+                param
+                for param in sig.parameters.values()
+                if param.kind in positional_args and param.default is param.empty
+            ]
+        )
+        if n_args != len(args_units):
+            raise TypeError(
+                f"The `expects` decorator used expects {len(args_units)} arguments, but a function expecting {n_args} "
+                f"arguments was wrapped"
+            )
```

Comment:
I would propose to change this to:
```suggestion
        params = sig.parameters

        bound = sig.bind_partial(*args_units, **kwargs_units)
        arguments = bound.arguments

        missing_params = {name for name in params.keys() if name not in arguments} - {"args", "kwargs"}
        if missing_params:
            raise TypeError(f"no units for {missing_params} found")
```